x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/action/EnrichCoordinatorProxyAction.java

```diff
@@ -239,12 +239,22 @@ static BiConsumer<MultiSearchRequest, BiConsumer<MultiSearchResponse, Exception>
                     ActionListener<MultiSearchResponse> listener = ActionListener.wrap(response -> {
                         shardResponses.put(enrichIndexName, new Tuple<>(response, null));
```

Comment:
This is tricky to me, `shardResponses` is a map outside the listener and keeps a copy of the `MultiSearchResponse`. This happens `itemsPerIndex` times. Each time, the listener "exits" with the map keeping a reference.

Then once all the searches are completed, it is reduced, potentially creating a NEW multisearchresponse with references to the response items in the individually collected responses.

I wonder if `shardResponses` should increment the references when storing the values, and then once this is all completed, decrement them.

Reply:
Hmm you're right there's a massive testing gap here in the enrich codebase. We only ever run with the counter going up to `1` so we just never hit this. It's definitely a bug, I'll address enhancing the tests first in a separate PR tomorrow and then we can get back to this, lets not mix this stuff :)

Reply:
https://github.com/elastic/elasticsearch/pull/102751 once that's in we should see this issue reproduce here

Reply:
Alright, I merged the above PR in here and it reproduced this issue. I fixed it just the way you describe above and it should be fine now.  If you're fine with this now and reviewing the other PR takes time, I can just back the test out of this PR before merging :)